File: frontend/src/services/userService.ts
import api from './api'

export interface User {
  id: number
  username: string
  email: string
  displayName?: string
  bio?: string
  role: string
  createdAt: string
}

export interface RegisterRequest {
  username: string
  email: string
  password: string
  displayName?: string
}

export interface LoginRequest {
  username: string
  password: string
}

export interface UpdateUserRequest {
  email?: string
  displayName?: string
  bio?: string
  password?: string
}

export interface AuthResponse {
  user: User
  token: string
}

export const userService = {
  register: (data: RegisterRequest) => api.post<User>('/auth/register', data),

  login: (data: LoginRequest) => api.post<AuthResponse>('/auth/login', data),

  getProfile: () => api.get<User>('/users/me'),

  getById: (id: number) => api.get<User>(`/users/${id}`),

  getByUsername: (username: string) => api.get<User>(`/users/username/${username}`),

  getAll: () => api.get<User[]>('/users'),

  update: (id: number, data: UpdateUserRequest) => api.put<User>(`/users/${id}`, data),

  delete: (id: number) => api.delete(`/users/${id}`),
}

File: backend/src/main/java/com/traceflow/service/UserService.java
package com.traceflow.service;

import com.traceflow.dto.UserDto;
import com.traceflow.exception.BadRequestException;
import com.traceflow.exception.ResourceNotFoundException;
import com.traceflow.model.User;
import com.traceflow.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @Transactional
    public UserDto.Response register(UserDto.RegisterRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new BadRequestException("Username already exists");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email already exists");
        }

        User user = User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .displayName(request.getDisplayName())
                .role(User.Role.NORMAL)
                .build();

        User saved = userRepository.save(user);
        return toResponse(saved);
    }

    public UserDto.Response getById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        return toResponse(user);
    }

    public UserDto.Response getByUsername(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("User", "username", username));
        return toResponse(user);
    }

    public List<UserDto.Response> getAll() {
        return userRepository.findAll().stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public UserDto.Response update(Long id, UserDto.UpdateRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));

        if (request.getEmail() != null) user.setEmail(request.getEmail());
        if (request.getDisplayName() != null) user.setDisplayName(request.getDisplayName());
        if (request.getBio() != null) user.setBio(request.getBio());
        if (request.getPassword() != null) user.setPassword(passwordEncoder.encode(request.getPassword()));

        User saved = userRepository.save(user);
        return toResponse(saved);
    }

    @Transactional
    public void delete(Long id) {
        userRepository.deleteById(id);
    }

    private UserDto.Response toResponse(User user) {
        return UserDto.Response.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .displayName(user.getDisplayName())
                .bio(user.getBio())
                .role(user.getRole().name())
                .createdAt(user.getCreatedAt() != null ? user.getCreatedAt().toString() : null)
                .build();
    }
}
